### PR TITLE
Fred/autocannon

### PR DESCRIPTION
**Pull request by Fredkim21**
LoadTest.ts is basic function for load testing, response data is the total send, total received, total missed, and the average response times that was received. The added some input locations but requires styling and better functionality. At the moment the URL needs to get linked to the original url location and need to connect the response to an actual chart. 
was not able to get any testing libraries to work in a node/typescript/electron setting. Artillery and k6 both require some command line functionality to work correctly and autocannon relies on node.js core modules which are also not available as an electron app. would have had to go about setting up another api endpoint in the serverside and runs the autocannon tests but that is still not certain to work so I just wrote a basic test.

### CODE REVIEW

**Review by csuzukida — APPROVED** [review]
LGTM, let us know if the # of requests getting sent off bug gets addressed

File: src/client/components/legacy-components/LoadTest.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        totalReceived += 1;
+        totalResponseTime += endTime - startTime;
+      } else {
+        // If the response has a non-success status, increment the totalMissed counter.
+        totalMissed += 1;
+      }
+    } catch (error) {
+      // If there is an error in sending the request, increment the totalMissed counter.
+      totalMissed += 1;
+    }
```
Comment (nattiechan) [inline src/client/components/legacy-components/LoadTest.ts:61]: What kind of error are we anticipating here? If it is error on Swell's end, should we count this as `missed` as well?

**Review by nattiechan — CHANGES_REQUESTED** [review]
Could we also put in a check to ensure only HTTP/2 requests (for now?) can show up in `ScheduleContainer.jsx`?

Also, I think we may want to rename the container file since this is no longer about scheduling requests

**Comment by Fredkim21** [comment]
in the bug fixes i fix the total number of requests getting sent and not being sent is getting recorded and returned. changed the name of the ScheduleContainer.jsx to TestContainer.tsx

File: src/client/components/legacy-components/LoadTest.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      // Wait for the calculated delay between requests before sending the next one.
+      await new Promise((resolve) => setTimeout(resolve, delayBetweenRequests));
+    }
+
+    // Check if the desired number of requests were sent
+    const desiredRequests = requestsPerSecond;
+    const actualRequests = requestsThisSecond;
+    if (actualRequests < desiredRequests) {
+      totalNotSent += desiredRequests - actualRequests;
+    }
```
Comment (nattiechan) [inline src/client/components/legacy-components/LoadTest.ts:92]: A few questions regarding this approach. I don't have the answer for any of these but curious what your thoughts are.

1. What does `totalNotSent` metric tell us? From the way this is setup it feels like it is accounting for the amount of time that Swell needs to set up and send the request, correct? Is this more of a reflection of the performance of Swell vs. the endpoint?

2. It looks like we are awaiting the request to be done before initiating the next one. What if the time it takes to send the request is longer than the request per second? Could we possibly utilize `setInterval` here, even with concurrent users?

File: src/client/components/legacy-components/TestContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        </div>
+      </div>
+      <div className="m-1">
+        <ReqResContainer displaySchedule={false} />
+      </div>
+      {runScheduledTests && (
+        <ScheduleReqResContainer scheduleInterval={scheduleInterval} />
+      )}
+      {!runScheduledTests && <StoppedContainer />}
+    </div>
```
Comment (nattiechan) [inline src/client/components/legacy-components/TestContainer.tsx:118]: Bump on my earlier comment regarding restricting the results to only HTTP/2 requests. I am wondering if this is incorporated somewhere?

**Review by nattiechan — COMMENTED** [review]
Looking good so far!

File: src/client/components/legacy-components/LoadTest.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+ * @param {number} requestsPerSecond - The number of requests per second each user should send.
+ * @param {number} durationInSeconds - The duration of the test in seconds.
+ *
+ * @returns {Promise<Object>} An object containing the load test results, including total sent requests,
+ *                            total received responses, total missed responses, and the average response time.
+ *
+ * @example
+ * const testResults = await simpleLoadTest('https://api.example.com/data', 10, 5, 30);
+ * console.log(testResults);
+ */
```
[the remark this answers is not in the document]
Comment (Fredkim21) [reply src/client/components/legacy-components/LoadTest.ts:15]: Wasnt sure on what we wanted to return yet so at the moment its fluid. but could add some more.

File: src/client/components/legacy-components/LoadTest.ts (hunk 2 above)
Comment (Fredkim21) [reply src/client/components/legacy-components/LoadTest.ts:92]: TotalNotSent, the way im thinking about it is the amount that the endpoint is not allowing in. So this is the case in terms of in production api endpoints.
Actually after some more testing it seems that this is still the case even on severs that I am working on in dev. I had gotten some limits even then. But this really seems to depend on the website and how they are monitoring and allowing refresh rates. Some websites only allow < 5 per calls per second where others are allowing the max, which seems to be about 70-80 per second.

File: src/client/components/legacy-components/LoadTest.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  totalReceived: number;
+  totalMissed: number;
+  averageResponseTime: number;
+  totalNotSent: number;
+  errorCounts: { [errorCode: string]: number };
+}
+
+export async function simpleLoadTest(
+  url: string,
+  // concurrentUsers: number, at this time concurrentUsers is not
```
Comment (nattiechan) [inline src/client/components/legacy-components/LoadTest.ts:27]: Let's remove this as well

File: src/client/components/legacy-components/LoadTest.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      }
+    }
+  };
+
+  // Define the runUserLoad function, which simulates a single user sending requests to the target URL.
+  const runUserLoad = async () => {
+    // Keep sending requests until the end time of the load test is reached.
+    while (Date.now() < endTest) {
+      const startOfSecond = Date.now();
+      let requestsThisSecond = 0;
```
Comment (nattiechan) [inline src/client/components/legacy-components/LoadTest.ts:79]: Let's define typing here

File: src/client/components/legacy-components/TestContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+              isDark ? '' : 'is-outlined'
+            } button-padding-vertical button-hover-color ml-3`}
+            onClick={async () => {
+              const results = await simpleLoadTest(
+                userUrl,
+                callsPerSecond,
+                totalTime
+              );
+              console.log(results);
+            }}
```
Comment (nattiechan) [inline src/client/components/legacy-components/TestContainer.tsx:78]: Looks like right now we are just logging the results in console logs. Have we thought about the plan on how to display the results in the UI? It is fine if we have not gotten that far yet, though I just want to check in where your thought process is at

**Review by nattiechan — COMMENTED** [review]
One last question before approval

File: src/client/components/legacy-components/LoadTest.ts (hunk 5 above)
Comment (Fredkim21) [reply src/client/components/legacy-components/LoadTest.ts]: ok removed

File: src/client/components/legacy-components/LoadTest.ts (hunk 6 above)
Comment (Fredkim21) [reply src/client/components/legacy-components/LoadTest.ts]: defined all the types i could find

File: src/client/components/legacy-components/TestContainer.tsx (hunk 7 above)
Comment (Fredkim21) [reply src/client/components/legacy-components/TestContainer.tsx:78]: I was planning on sending the results in console.log into the response body but just havent been able to get to it at the moment. Am open to ideas on to where and how data should be displayed.

**Review by nattiechan — APPROVED** [review]
Looks great! This is contingent on the consensus that the UI will be done in a separate PR :)

**Pull request merged**